File: src/handlers/pages.rs
use crate::dev::{SCRIPT, WS_PORT};
use crate::error::{errors_to_html, ErrorType, MapProcErr, ProcessError, WithItem};
use crate::handlers::components::{process_component, ComponentTypes};
use crate::handlers::katex_assets;
use crate::handlers::markdown::render_markdown;
use crate::handlers::templates::process_template;
use crate::utils::ProcessResult;
use crate::IS_DEV;
use minify_html::minify;
use rayon::prelude::*;
use std::sync::Arc;
use std::{collections::HashSet, fs, path::PathBuf};

fn process_step<F>(
    func: F,
    src: &PathBuf,
    string: &mut String,
    hist: &HashSet<PathBuf>,
    vec_errs: &mut Vec<ProcessError>,
) where
    F: Fn(&PathBuf, String, &HashSet<PathBuf>) -> ProcessResult,
{
    let result = func(src, std::mem::take(string), hist);
    *string = result.output;
    vec_errs.extend(result.errors);
}

pub fn page(src: &PathBuf, mut string: String, hist: HashSet<PathBuf>) -> ProcessResult {
    let mut errors: Vec<ProcessError> = Vec::new();

    if string.contains("</markdown>") {
        let md_result = render_markdown(string);
        string = md_result.output;
        errors.extend(md_result.errors);
    }

    process_step(
        |srcpath, str, hist| {
            process_component(srcpath, str, ComponentTypes::Wrapping, hist.clone())
        },
        src,
        &mut string,
        &hist,
        &mut errors,
    );
    process_step(
        |srcpath, str, hist| {
            process_component(srcpath, str, ComponentTypes::SelfClosing, hist.clone())
        },
        src,
        &mut string,
        &hist,
        &mut errors,
    );
    process_step(
        |srcpath, str, hist| process_template(srcpath, str, hist.clone()),
        src,
        &mut string,
        &hist,
        &mut errors,
    );

    ProcessResult {
        output: string,
        errors,
    }
}

pub fn process_pages(
    dir: &PathBuf,
    src: &PathBuf,
    source: PathBuf,
    pages: PathBuf,
) -> Result<(), Vec<ProcessError>> {
    let mut errors: Vec<ProcessError> = Vec::new();
    let dev = *IS_DEV.get().unwrap();

    let entries = match fs::read_dir(&pages) {
        Ok(entries) => entries,
        Err(e) => {
            errors.push(ProcessError {
                error_type: ErrorType::Io,
                item: WithItem::File,
                path: pages.clone(),
                message: Some(format!("Error reading pages directory: {:?}", e)),
            });
            return Err(errors);
        }
    };

    let working_dir = if dev { "dev" } else { "dist" };
    let minify_cfg = Arc::new(minify_html::Cfg::new());

    let mut file_tasks = Vec::new();
    let mut dir_tasks = Vec::new();

    for entry in entries {
        match entry {
            Ok(entry) => {
                let path = entry.path();
                if path.is_dir() {
                    dir_tasks.push((dir.clone(), src.clone(), source.join(&path), path));
                } else {
                    file_tasks.push(path);
                }
            }
            Err(e) => {
                errors.push(ProcessError {
                    error_type: ErrorType::Io,
                    item: WithItem::File,
                    path: pages.clone(),
                    message: Some(format!("Failed to read directory entry: {}", e)),
                });
            }
        }
    }

    // Process directories sequentially
    for (dir, src, source, path) in dir_tasks {
        if let Err(mut errs) = process_pages(&dir, &src, source, path) {
            errors.append(&mut errs);
        }
    }

    // Process files in parallel using rayon
    if !file_tasks.is_empty() {
        let results: Vec<(Vec<ProcessError>,)> = file_tasks
            .par_iter()
            .map(|path| {
                let errs = process_single_file(
                    path.clone(),
                    dir.clone(),
                    src.clone(),
                    working_dir.to_string(),
                    dev,
                    Arc::clone(&minify_cfg),
                );
                (errs,)
            })
            .collect();

        for (errs,) in results {
            errors.extend(errs);
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

/// Process a single page file. Always returns collected errors (may be empty).
/// In dev mode, writes an error page if there are errors.
/// In build mode, skips writing the file if there are errors.
fn process_single_file(
    path: PathBuf,
    dir: PathBuf,
    src: PathBuf,
    working_dir: String,
    dev: bool,
    minify_cfg: Arc<minify_html::Cfg>,
) -> Vec<ProcessError> {
    let mut errors: Vec<ProcessError> = Vec::new();

    // Reset KaTeX usage flag for this page
    katex_assets::reset_katex_flag();

    let file_content =
        match fs::read_to_string(&path).map_proc_err(WithItem::File, ErrorType::Io, &path, None) {
            Ok(content) => content,
            Err(e) => {
                errors.push(e);
                write_error_page_if_dev(dev, &errors, &dir, &src, &path, &working_dir);
                return errors;
            }
        };

    if file_content.is_empty() {
        errors.push(ProcessError {
            error_type: ErrorType::Other,
            item: WithItem::File,
            path: path.clone(),
            message: Some("Page file is empty".to_string()),
        });
        write_error_page_if_dev(dev, &errors, &dir, &src, &path, &working_dir);
        return errors;
    }

    let result = page(&src, file_content, HashSet::new());
    errors.extend(result.errors);

    let out_path = match resolve_out_path(&path, &dir, &src, &working_dir) {
        Ok(p) => p,
        Err(e) => {
            errors.push(e);
            return errors;
        }
    };

    if let Some(parent) = out_path.parent() {
        if let Err(e) = fs::create_dir_all(parent) {
            errors.push(ProcessError {
                error_type: ErrorType::Io,
                item: WithItem::File,
                path: out_path.clone(),
                message: Some(format!("Failed to create directory: {}", e)),
            });
            return errors;
        }
    }

    // If there are errors: dev → error page, build → skip
    if !errors.is_empty() {
        if dev {
            let dev_script = make_dev_script();
            let error_html = errors_to_html(&errors, dev_script.as_deref());
            let _ = fs::write(&out_path, error_html.as_bytes());
        }
        return errors;
    }

    let mut output = result.output;

    // Inject KaTeX CSS if math was rendered (unless disabled)
    if katex_assets::was_katex_used() && !katex_assets::is_katex_injection_disabled() {
        katex_assets::print_katex_message();

        if output.contains("<head>") {
            output = output.replace(
                "<head>",
                &format!("<head>\n{}", katex_assets::get_katex_css_tag()),
            );
        } else {
            output = format!("{}\n{}", katex_assets::get_katex_css_tag(), output);
        }
    }

    if dev {
        let ws_port = *WS_PORT.get().unwrap();
        if !output.contains("// * SCRIPT INCLUDED IN DEV MODE") {
            output = output.replace("<head>", &format!("<head>{}", SCRIPT));
            output = output.replace("__SIMPLE_WS_PORT_PLACEHOLDER__", &ws_port.to_string());
        }
    }

    let to_write = if dev {
        output.into_bytes()
    } else {
        minify(output.as_bytes(), &minify_cfg)
    };

    if let Err(e) = fs::write(&out_path, &to_write) {
        errors.push(ProcessError {
            error_type: ErrorType::Io,
            item: WithItem::File,
            path: out_path,
            message: Some(format!("Failed to write file: {}", e)),
        });
    }

    errors
}

fn resolve_out_path(
    path: &PathBuf,
    dir: &PathBuf,
    src: &PathBuf,
    working_dir: &str,
) -> Result<PathBuf, ProcessError> {
    let relative_to_src = path.strip_prefix(src).map_err(|e| ProcessError {
        error_type: ErrorType::Io,
        item: WithItem::File,
        path: path.clone(),
        message: Some(format!("Failed to strip src prefix: {}", e)),
    })?;

    let relative_to_pages = relative_to_src
        .strip_prefix("pages")
        .map_err(|e| ProcessError {
            error_type: ErrorType::Io,
            item: WithItem::File,
            path: path.clone(),
            message: Some(format!("Failed to strip pages prefix: {}", e)),
        })?;

    Ok(dir.join(working_dir).join(relative_to_pages))
}

fn make_dev_script() -> Option<String> {
    let ws_port = WS_PORT.get()?;
    Some(SCRIPT.replace("__SIMPLE_WS_PORT_PLACEHOLDER__", &ws_port.to_string()))
}

/// In dev mode, attempt to write an error page for the given source path.
/// Best-effort — if path resolution fails, this is a no-op.
fn write_error_page_if_dev(
    dev: bool,
    errors: &[ProcessError],
    dir: &PathBuf,
    src: &PathBuf,
    path: &PathBuf,
    working_dir: &str,
) {
    if !dev {
        return;
    }
    if let Ok(out_path) = resolve_out_path(path, dir, src, working_dir) {
        if let Some(parent) = out_path.parent() {
            let _ = fs::create_dir_all(parent);
        }
        let dev_script = make_dev_script();
        let error_html = errors_to_html(errors, dev_script.as_deref());
        let _ = fs::write(&out_path, error_html.as_bytes());
    }
}

File: src/handlers/pages.cpp
#include "handlers/pages.hpp"
#include "handlers/components.hpp"
#include "handlers/katex_assets.hpp"
#include "handlers/markdown.hpp"
#include "handlers/templates.hpp"
#include "minify.hpp"
#include <fstream>
#include <mutex>
#include <sstream>
#include <thread>
#include <vector>

namespace simple::handlers {

extern bool is_dev;
extern uint16_t ws_port;
extern std::string_view script;

auto page(const std::filesystem::path &src, std::string string,
          std::unordered_set<std::filesystem::path> hist) -> ProcessResult {

  std::vector<ProcessError> errors;

  if (string.find("</markdown>") != std::string::npos) {
    string = render_markdown(std::move(string));
  }

  auto wrapping_result =
      process_component(src, std::move(string), ComponentTypes::Wrapping, hist);
  string = std::move(wrapping_result.output);
  errors.insert(errors.end(), wrapping_result.errors.begin(),
                wrapping_result.errors.end());

  auto self_closing_result = process_component(
      src, std::move(string), ComponentTypes::SelfClosing, hist);
  string = std::move(self_closing_result.output);
  errors.insert(errors.end(), self_closing_result.errors.begin(),
                self_closing_result.errors.end());

  auto template_result = process_template(src, std::move(string), hist);
  string = std::move(template_result.output);
  errors.insert(errors.end(), template_result.errors.begin(),
                template_result.errors.end());

  return ProcessResult{std::move(string), std::move(errors)};
}

auto process_single_file(const std::filesystem::path &path,
                         const std::filesystem::path &dir,
                         const std::filesystem::path &src,
                         std::string_view working_dir) -> Results<void> {

  std::vector<ProcessError> errors;

  reset_katex_flag();

  std::ifstream file{path};
  if (!file) {
    return std::unexpected(std::vector{ProcessError{
        ErrorType::Io, WithItem::File, path, "Failed to read file"}});
  }

  std::stringstream buffer;
  buffer << file.rdbuf();
  auto file_content = buffer.str();

  if (file_content.empty()) {
    return std::unexpected(std::move(errors));
  }

  auto result = page(src, std::move(file_content), {});
  errors.insert(errors.end(), result.errors.begin(), result.errors.end());

  auto relative_to_src = std::filesystem::relative(path, src);
  auto relative_to_pages = std::filesystem::relative(relative_to_src, "pages");
  auto out_path = dir / working_dir / relative_to_pages;

  if (out_path.has_parent_path()) {
    std::error_code ec;
    std::filesystem::create_directories(out_path.parent_path(), ec);
    if (ec) {
      errors.push_back(ProcessError{
          ErrorType::Io, WithItem::File, out_path,
          std::format("Failed to create directory: {}", ec.message())});
      return std::unexpected(std::move(errors));
    }
  }

  auto output = std::move(result.output);

  if (was_katex_used() && !is_katex_injection_disabled()) {
    print_katex_message();

    auto katex_css = get_katex_css_tag();
    if (output.find("<head>") != std::string::npos) {
      auto head_tag = std::format("<head>\n{}", katex_css);
      size_t pos = output.find("<head>");
      output.replace(pos, 6, head_tag);
    } else {
      output = std::format("{}\n{}", katex_css, output);
    }
  }

  if (is_dev) {
    if (output.find("// * SCRIPT INCLUDED IN DEV MODE") == std::string::npos) {
      auto head_with_script = std::format("<head>{}", script);
      size_t pos = output.find("<head>");
      if (pos != std::string::npos) {
        output.replace(pos, 6, head_with_script);
      }

      size_t ws_pos = 0;
      while ((ws_pos = output.find("__SIMPLE_WS_PORT_PLACEHOLDER__", ws_pos)) !=
             std::string::npos) {
        output.replace(ws_pos, 30, std::to_string(ws_port));
        ws_pos += std::to_string(ws_port).length();
      }
    }
  }

  std::string to_write;
  if (is_dev) {
    to_write = std::move(output);
  } else {
    to_write = minify_html(output);
  }

  std::ofstream out_file{out_path, std::ios::binary};
  if (!out_file) {
    errors.push_back(ProcessError{ErrorType::Io, WithItem::File, out_path,
                                  "Failed to write file"});
    return std::unexpected(std::move(errors));
  }

  out_file << to_write;

  if (errors.empty()) {
    return {};
  }
  return std::unexpected(std::move(errors));
}

auto process_pages(const std::filesystem::path &dir,
                   const std::filesystem::path &src,
                   const std::filesystem::path &source,
                   const std::filesystem::path &pages) -> Results<void> {

  std::vector<ProcessError> errors;
  std::error_code ec;

  std::string working_dir = is_dev ? "dev" : "dist";

  std::vector<std::filesystem::path> file_tasks;
  std::vector<std::tuple<std::filesystem::path, std::filesystem::path,
                         std::filesystem::path, std::filesystem::path>>
      dir_tasks;

  for (const auto &entry : std::filesystem::directory_iterator(pages, ec)) {
    if (ec) {
      errors.push_back(ProcessError{
          ErrorType::Io, WithItem::File, pages,
          std::format("Error reading pages directory: {}", ec.message())});
      return std::unexpected(std::move(errors));
    }

    auto path = entry.path();
    if (entry.is_directory()) {
      dir_tasks.emplace_back(dir, src, source / path, path);
    } else {
      file_tasks.push_back(path);
    }
  }

  for (const auto &[d, s, src_path, p] : dir_tasks) {
    auto result = process_pages(d, s, src_path, p);
    if (!result) {
      errors.insert(errors.end(), result.error().begin(), result.error().end());
    }
  }

  std::mutex errors_mutex;
  std::vector<std::thread> threads;
  auto thread_count = std::min(std::thread::hardware_concurrency(),
                               static_cast<unsigned int>(file_tasks.size()));

  if (thread_count == 0)
    thread_count = 1;

  size_t chunk_size = (file_tasks.size() + thread_count - 1) / thread_count;

  for (size_t i = 0; i < thread_count; ++i) {
    threads.emplace_back([&, i, chunk_size]() {
      size_t start = i * chunk_size;
      size_t end = std::min(start + chunk_size, file_tasks.size());

      for (size_t j = start; j < end; ++j) {
        auto result = process_single_file(file_tasks[j], dir, src, working_dir);
        if (!result) {
          std::lock_guard lock{errors_mutex};
          errors.insert(errors.end(), result.error().begin(),
                        result.error().end());
        }
      }
    });
  }

  for (auto &thread : threads) {
    thread.join();
  }

  if (errors.empty()) {
    return {};
  }
  return std::unexpected(std::move(errors));
}

} // namespace simple::handlers
